qep-401-overview-widget-types.md

```diff
@@ -0,0 +1,74 @@
+# QGIS Enhancement: Overview of widget types, constraints and default values
+
+**Date** 2025/01/07
+
+**Author** Germán Carrillo (@gacarrillor), Mathieu Pellerin (@nirvn)
+
+**Contact** german at opengis dot ch, mathieu at opengis dot ch
+
+**Version** QGIS 3.4X.X+
+
+# Summary
+
+
+In the vector layer properties, QGIS allows users to drag available layer fields from a field panel and drop them in a form layout panel. Once there, each form layout item has a widget type assigned with a default configuration. These widget types are currently only shown individually in a third panel, where users can change both the widget type and its configuration.
+
+![Screenshot from 2024-11-24 22-22-18](https://github.com/user-attachments/assets/47cee69d-6ff2-4f50-9069-eab44bb66757)
+
+When a layer has a relatively large number of fields, there is a need to have an overview of widget types, rather than just seeing them one by one. This can be easily extended to other field settings like constraints and default values.
+
+We propose to tackle this via visual hints in field and form item panels, as well as bringing handy stuff to improve user experience.
+
+## Proposed Solution
+
+Switch from current item-based tree widget to model-based tree view to ease maintenance and gain flexibility.
+
+Use the form layout item’s decoration role (i.e., item’s icon) to show its type of widget. This is appropriate because all items will have one and only one type of widget.
+
+Use right-aligned indicator icons (like those that are used in the QGIS layer tree for filters and missing data sources, among others) for constraints and default values. The indicators will only be shown when constraints and/or default values are present, and details about specific configurations will be displayed via tooltips.
+
+This work includes the graphic design of almost 20 icons for widget types, as well as a couple of icons for indicators of constraints and default values.
+
+
+#### Other UI/UX enhancements
+
+ + **Filter/search box**, taking advantage of the model-based tree view.
+
+ + **Form preview**, to give users the option to see what they'll get easily and in the right place, i.e., without having to follow additional steps like closing Layer properties, enabling editing and adding a feature just to see what they've just configured in action. The preview will be opened in a new window, parented to the Vector Layer Properties dialog.
```

Comment:
Instead of having another windows preview, could we not just update the real form when user clicks on *Apply* button ? It's actually the case when you change the widget, but not when we change the alias for instance.

If it was working for any parameter you can modify, it would solve the preview issue in an even better way, because it would be populated with real data.

Reply:
We haven't really aimed at changing existing QGIS behavior on opened forms. The form preview will only be shown as a child window of the Vector Layer Properties dialog. Users could preview the configured form without leaving that dialog, i.e., without having to open the attribute table's form view or to add a new geometry.

Reply:
I understand but it's just yet another window (and more code) when there is actually one existing, already populated with relevant data so you can better look about what it really look like.

Do you plan to populate this preview window with a feature like what we are doing for expression evaluation for instance, or it's more like the Qt designer empty preview?

Reply:
Fair question about how the form will be populated with data.

Regarding the proposal to reuse the "real" form, I wonder if that's a good experience. The layer properties dialog (where widgets/forms are configured) is a modal dialog, so it won't be possible to open a form through the main window and already open forms will be stacked behind the layer properties and not interactive. I am not sure we can realistically reuse these.

Reply:
Fair point too... I tested it before posting my comment but I was actually using dual screen.

I like current behavior with dual screen though
![widgets](https://github.com/user-attachments/assets/af04cb2e-f1f4-4314-86ec-d33d12f20b09)

here we see that form is not refresh when alias is modified, but only when widget type is.

We could also do like we do with map symbology and display widget editor beside the attribute form when hitting F7 and refresh the form as soon as user modify a parameter.

Reply:
> Do you plan to populate this preview window with a feature like what we are doing for expression evaluation for instance, or it's more like the Qt designer empty preview?

I guess I always had this Qt-designer's empty preview in mind (i.e., that would already be great), but adding sample data to the preview in the same way that it is done for previewing expressions results sounds awesome. 

Of course, we won't have a `Save` button available there, but the rest could look like a real input form. That is, widgets wouldn't be set to read-only mode (unless the widget config says so), but any changes to the sample data wouldn't be applied.

Reply:
> Do you plan to populate this preview window with a feature like what we are doing for expression evaluation for instance, or it's more like the Qt designer empty preview?

Suggestion applied in https://github.com/qgis/QGIS-Enhancement-Proposals/pull/318/commits/bcd86d40a150575c78556a434ad4ce42ca58c16c, thanks!

> We could also do like we do with map symbology and display widget editor beside the attribute form when hitting F7 and refresh the form as soon as user modify a parameter.

With respect to this, it looks to be in-line with what @saberraz suggested for future iterations.
We'd consider moving the UI to a docked panel a bit out of scope for now.